pynobo.py

```diff
@@ -145,205 +148,326 @@ class API:
+            self.transport = transport
+
+        def datagram_received(self, data: bytes, addr):
+            msg = data.decode('utf-8')
+            _LOGGER.info('broadcast received: %s from %s', msg, addr[0])
+            # Expected string “__NOBOHUB__123123123”, where 123123123 is replaced with the first 9 digits of the Hub’s serial number.
+            if msg.startswith('__NOBOHUB__'):
+                discover_serial = msg[11:]
+                discover_ip = addr[0]
+                if len(self.serial) == 12:
+                    if discover_serial != self.serial[0:9]:
+                        # This is not the Ecohub you are looking for
+                        discover_serial = None
+                    else:
+                        discover_serial = self.serial
+                else:
+                    discover_serial += self.serial
+                if self.ip and discover_ip != self.ip:
+                    # This is not the Ecohub you are looking for
+                    discover_ip = None
+                if discover_ip and discover_serial:
+                    self.hubs.add( (discover_ip, discover_serial) )
+
+    def __init__(self, serial, ip=None, discover=True, loop: asyncio.AbstractEventLoop = None):
+        """
+        Initialize logger and dictionaries.
 
-        Keyword arguments:  
-        serial -- The last 3 digits of the Ecohub serial number or the complete 12 digit serial number  
-        ip -- ip address to search for Ecohub at (default None)  
-        discover -- True/false for using UDP autodiscovery for the IP (default True)
+        :param serial: The last 3 digits of the Ecohub serial number or the complete 12 digit serial number
+        :param ip: ip address to search for Ecohub at (default None)
+        :param discover: True/false for using UDP autodiscovery for the IP (default True)
+        :param loop: the asyncio event loop to use
         """
-        self.logger = logging.getLogger(__name__)
+
+        self._loop = asyncio.get_event_loop() if loop is None else loop
+        self.serial = serial
+        self.ip = ip
+        self.discover = discover
+
+        self._callbacks = []
+        self._reader = None
+        self._writer = None
+        self._keep_alive_task = None
+        self._socket_receive_task = None
+
+        self._received_all_info = False
         self.hub_info = {}
         self.zones = collections.OrderedDict()
         self.components = collections.OrderedDict()
         self.week_profiles = collections.OrderedDict()
         self.overrides = collections.OrderedDict()
         self.temperatures = collections.OrderedDict()
 
+        # Start asyncio in a separate thread unless an event loop was provided.
+        if loop is None:
+            self._loop.run_until_complete(self.start())
+            self._thread = threading.Thread(target=lambda: self._loop.run_forever())
+            self._thread.start()
```

Comment:
I think we should do a 
`            self._thread.setDaemon(True)`
before the `self._thread.start()`. If you do not specify an event loop, I do not think you expect this thread to keep your program running after the main thread exists either.

Reply:
Yes. Done.